File: src/main/java/remake/internal/Log.java
package remake.internal;

import arc.files.Fi;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.MessageFormat;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import static essentials.Global.getTime;
import static essentials.Global.nbundle;
import static essentials.Main.root;
import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;

public class Log {
    final static arc.util.Log.DefaultLogHandler handler = new arc.util.Log.DefaultLogHandler();

    public static enum LogType {
        log, warn, error, debug, server, serverwarn, servererr, client, clientwarn, clienterr, config, player, playerwarn, playererr,
        tap, withdraw, block, deposit, chat, griefer, web
    }

    public static void info(String value, Object... parameter) {
        String result = new Bundle().get(value);
        if (result != null) {
            handler.log(arc.util.Log.LogLevel.info, MessageFormat.format(result, parameter));
        } else {
            handler.log(arc.util.Log.LogLevel.info, value);
        }
    }

    public static void err(String value, Object... parameter) {
        String result = new Bundle().get(value);
        if (result != null) {
            handler.log(arc.util.Log.LogLevel.err, MessageFormat.format(result, parameter));
        } else {
            handler.log(arc.util.Log.LogLevel.err, value);
        }
    }

    public static void warn(String value, Object... parameter) {
        String result = new Bundle().get(value);
        if (result != null) {
            handler.log(arc.util.Log.LogLevel.warn, MessageFormat.format(result, parameter));
        } else {
            handler.log(arc.util.Log.LogLevel.warn, value);
        }
    }

    public static void write(LogType type, String value) {
        String date = DateTimeFormatter.ofPattern("yyyy-MM-dd HH_mm_ss").format(LocalDateTime.now());
        Path newlog = Paths.get(root.child("log/" + type + ".log").path());
        Path oldlog = Paths.get(root.child("log/old/" + type + "/" + date + ".log").path());
        Fi mainlog = root.child("log/" + type + ".log");
        Fi logfolder = root.child("log");

        if (mainlog != null && mainlog.length() > 1024 * 256) {
            mainlog.writeString(nbundle("log-file-end", date), true);
            try {
                Files.move(newlog, oldlog, REPLACE_EXISTING);
            } catch (IOException e) {
                new CrashReport(e);
            }
            mainlog = null;
        }

        if (mainlog == null) {
            mainlog = logfolder.child(type + ".log");
        }

        mainlog.writeString("[" + getTime() + "]" + value + "\n", true);
    }
}

File: old/internal/Log.kt
package essentials.internal

import arc.util.Log
import essentials.Main.Companion.pluginRoot
import essentials.data.Config
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.text.MessageFormat
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

object Log {
    fun info(value: String, vararg parameter: Any?) {
        val result = Bundle()[value]
        Log.info(MessageFormat.format(result, *parameter))
    }

    fun err(value: String, vararg parameter: Any?) {
        val result = Bundle()[value]
        Log.err(MessageFormat.format(result, *parameter))
    }

    fun warn(value: String, vararg parameter: Any?) {
        val result = Bundle()[value]
        Log.warn(MessageFormat.format(result, *parameter))
    }

    fun server(value: String, vararg parameter: Any?) {
        val result = Bundle()[value]
        Log.info("[EssentialServer] " + MessageFormat.format(result, *parameter))
    }

    fun client(value: String, vararg parameter: Any?) {
        val result = Bundle()[value]
        Log.info("[EssentialClient] " + MessageFormat.format(result, *parameter))
    }

    fun player(value: String, vararg parameter: Any?) {
        val result = Bundle()[value]
        Log.info("[EssentialPlayer] " + MessageFormat.format(result, *parameter))
    }

    fun write(type: LogType, value: String, vararg params: String) {
        if(Config.logging) {
            val date = DateTimeFormatter.ofPattern("yyyy-MM-dd HH_mm_ss").format(LocalDateTime.now())
            val newlog = Paths.get(pluginRoot.child("log/$type.log").path())
            val oldlog = Paths.get(pluginRoot.child("log/old/$type/$date.log").path())
            var mainlog = pluginRoot.child("log/$type.log")
            val logfolder = pluginRoot.child("log")

            if(mainlog != null && mainlog.length() > 1024 * 256) {
                mainlog.writeString(Bundle()["log.file-end", date], true)
                try {
                    if(!pluginRoot.child("log/old/$type").exists()) {
                        pluginRoot.child("log/old/$type").mkdirs()
                    }
                    Files.move(newlog, oldlog, StandardCopyOption.REPLACE_EXISTING)
                } catch(e: IOException) {
                    CrashReport(e)
                }
                mainlog = null
            }
            if(mainlog == null) mainlog = logfolder.child("$type.log")
            mainlog!!.writeString("[${Tool.getLocalTime()}] ${Bundle().get(value, *params)}\n", true)
        }
    }

    enum class LogType {
        Log, Warn, Error, Debug, Server, ServerWarn, ServerError, Client, ClientWarn, ClientError, Config, Player, PlayerWarn, PlayerError, Tap, WithDraw, Block, Deposit, Chat, Griefer, Web
    }
}